osu.Game/Screens/Play/Player.cs

```diff
@@ -412,12 +412,22 @@ public void Restart()
 
         private ScheduledDelegate completionProgressDelegate;
 
-        private void onCompletion()
+        private void updateCompletionState(ValueChangedEvent<bool> completionState)
         {
             // screen may be in the exiting transition phase.
             if (!this.IsCurrentScreen())
                 return;
 
+            // cancel push delegate in case judges reverted
+            // after delegate may have already been scheduled.
+            if (!completionState.NewValue)
```

Comment:
using `OldValue` in bool events is redundant here as I haven't used `BindValueChanged(..., true)` which makes `NewValue always the opposite of `OldValue`, therefore not adding anything.

btw the operations done inside this block is required to make the cancellation correct (hopefully), as they all reset everything that was set when a completion occurs:
 - `completionProgressDelegate?.Cancel()` – cancel the completion delegate to prevent from pushing. the delegate may already be null, so add a null-conditional, due to `allowResults` being false, thus not creating the delegate.
 - `completionProgressDelegate = null` – reset the delegate back to null to assure the `Player` that there isn't a push delegate scheduled.
- `ValidForResume = true;` – Reset back `ValidForResume` from being false to allow pausing back (though just remembered that rewinding comes only on replays which you can't pause, but to ensure that there are no traces of the player going through the completion process.)

Reply:
Fair point about the old value being redundant, although there was no need to explain the rest as it was pretty obvious even to me what it did.

Reply:
ok great, *I was literally about to start commenting on each line in case something wasn't obvious*